hal/src/sercom/v2/uart/pads_thumbv6m.rs

```diff
@@ -0,0 +1,486 @@
+//! UART pad definitions for thumbv6m targets
+
+use core::marker::PhantomData;
+
+use super::{AnyConfig, Capability, CharSize, Config, Duplex, Rx, Tx};
+use crate::sercom::v2::*;
+use crate::typelevel::{NoneT, Sealed};
+
+#[cfg(not(feature = "samd11"))]
+use crate::gpio::v2::AnyPin;
+
+//=============================================================================
+// RxpoTxpo
+//=============================================================================
+
+/// Configure the `RXPO` and `TXPO` fields based on a set of [`Pads`]
+///
+/// According to the datasheet, the `RXPO` and `TXPO` values specify which
+/// SERCOM pads are used for various functions. Moreover, depending on which
+/// pads are actually in use, only certain combinations of these values make
+/// sense and are valid.
+///
+/// This trait is implemented for valid, four-tuple combinations of
+/// [`OptionalPadNum`]s. Those implementations are then lifted to the
+/// corresponding [`Pads`] types.
+///
+/// To satisfy this trait, the combination of [`OptionalPadNum`]s must specify
+/// [`PadNum`] for at least one of `RX` and `TX`. Furthermore, no
+/// two [`PadNum`]s can conflict.
+pub trait RxpoTxpo {
+    /// `RXPO` field value
+    const RXPO: u8;
+
+    /// `RXPO` field value
+    const TXPO: u8;
+}
+
+/// Lift the implementations of [`RxpoTxpo`] from four-tuples of
+/// [`OptionalPadNum`]s to the corresponding [`Pads`] types.
+impl<S, RX, TX, RTS, CTS> RxpoTxpo for Pads<S, RX, TX, RTS, CTS>
+where
+    S: Sercom,
+    RX: OptionalPad,
+    TX: OptionalPad,
+    RTS: OptionalPad,
+    CTS: OptionalPad,
+    (RX::PadNum, TX::PadNum, RTS::PadNum, CTS::PadNum): RxpoTxpo,
+{
+    const RXPO: u8 = <(RX::PadNum, TX::PadNum, RTS::PadNum, CTS::PadNum)>::RXPO;
+    const TXPO: u8 = <(RX::PadNum, TX::PadNum, RTS::PadNum, CTS::PadNum)>::TXPO;
+}
+
+//=============================================================================
+// Implement RxpoTxpo
+//=============================================================================
+
+/// Filter [`PadNum`] permutations and implement [`RxpoTxpo`]
+    #[rustfmt::skip]
+    macro_rules! impl_rxpotxpo {
```

Comment:
Does UART have loopback mode, like SPI? I recently modified this macro for SPI to avoid accidentally using loopback mode when one of the pads is `NoneT`. If that matters at all to you, see the updated version in #467.

Reply:
Out of curiosity, I looked at the SAMD21 datasheet and it doesn't explicitly say that the UART supports loopback like for SPI, but it appears that RXPO can be set to be set to any of the pads so it probably does support a loopback so long as you're not using hardware flow control.

Reply:
From the datasheet:
> 25.6.3.6 Loop-back Mode
By configuring the Receive Data Pinout (CTRLA.RXPO) and Transmit Data Pinout (CTRLA.TXPO) to use the same data
pins for transmit and receive, loop-back is achieved. The loop-back is through the pad, so the signal is also available
externally.

Reply:
This was a problem for me, because if you create an SPI with `Rx` `Capability`, you still need to send dummy data to initiate the transfer. But if you were accidentally in loop-back mode, you would be reading back that dummy data.

I'm not sure if that's an issue for you though. I guess not? If you're `Rx`, then you're never sending data, so you can't read it back. But would you accidentally have contention on the line? If the SAMD transmit pin is trying to hold the line high, for idle, but the sender on the other end is trying to bring it low, for a transfer, would you have a problem?

Reply:
Here is where I corrected it:

https://github.com/bradleyharden/atsamd/blob/3d9294b6d7f87b2fd5eae9d762f2bc73fca1b8b9/hal/src/sercom/v2/spi/pads_thumbv6m.rs#L103

Reply:
Well, I don't *think* it would create any problems. However I'd rather be proactive and avoid potential problems in the future, so I'm going to avoid entering loopback mode by only enabling `RXEN`/`TXEN`/Both depending on the `Capability`. Right now both unconditionally get enabled.